tmpfile/tmpfile.go

```diff
@@ -0,0 +1,147 @@
+package tmpfile
+
+import (
+	"fmt"
+	"io"
+	"io/ioutil"
+	"os"
+	"path/filepath"
+	"sync"
+
+	"github.com/odeke-em/go-uuid"
+)
+
+type readSeekerAt interface {
+	io.ReadSeeker
+	ReadAt(b []byte, off int64) (int, error)
+}
+
+type readSeekerAtWriteCloser interface {
+	readSeekerAt
+	io.WriteCloser
+}
+
+type statInfoer interface {
+	Stat() (os.FileInfo, error)
+}
+
+type readSeekerCloseWriterStatInfoer interface {
+	readSeekerAtWriteCloser
+	statInfoer
+}
+
+var _ readSeekerCloseWriterStatInfoer = &TmpFile{}
+
+type TmpFile struct {
+	tmpf          *os.File
+	dir           string
+	path          string
+	doneOnce      sync.Once
+	ownCreatedDir bool
+}
+
+func (tf TmpFile) Done() error {
+	var err error
+	tf.doneOnce.Do(func() {
+		_ = tf.tmpf.Close()
+		targetPath := tf.tmpf.Name()
+		if tf.ownCreatedDir {
+			// Since we are removing the entire dir,
+			// any file in the dir will be cleaned out.
+			targetPath = tf.dir
+		}
+		err = os.RemoveAll(targetPath)
+	})
+	return err
+}
+
+type Context struct {
+	Suffix            string
+	Dir               string
+	CreateIsolatedDir bool
+	// NoOverrideIfSuffixEmpty if set signifies that the caller does not want
+	// any suffix to be created for them if no suffix was previously provided,
+	// that is they want to main the empty suffix.
+	NoOverrideIfSuffixEmpty bool
+}
+
+func New(ctx *Context) (*TmpFile, error) {
+	if ctx == nil {
+		return nil, fmt.Errorf("nil context passed in")
+	}
+
+	dir := ctx.Dir
+	var abortFn func() error
+	ownCreatedDir := false
+	if ctx.CreateIsolatedDir {
+		dir = filepath.Join(os.Getenv("TMPDIR"), dir, uuid.UUID4().String())
```

Comment:
I think I'm having an issue from this. My system doesn't have TMPDIR. Maybe it could use os.TempDir? or ioutil.TempDir.